src/QueryAggregator.js

```diff
@@ -0,0 +1,109 @@
+import invariant from 'invariant';
+import Relay from 'react-relay';
+
+import getParamsForRoute from './getParamsForRoute';
+
+export default class QueryAggregator {
+  constructor() {
+    this.uniqueQueryNames = new WeakMap();
+    this.lastQueryIndex = 0;
+
+    this.data = {};
+  }
+
+  updateRoute(routerProps) {
+    const {params, location} = routerProps;
+
+    const relayRoute = {
+      queries: {},
+      params: {},
+    };
+    const fragmentSpecs = {};
+
+    routerProps.branch.forEach(route => {
+      const {queries} = route;
+      if (!queries) {
+        return;
+      }
+
+      const {component} = route;
+
+      // In principle not all container component routes have to specify
+      // queries, because some of them might somehow receive fragments from
+      // their parents, but it would definitely be wrong to specify queries
+      // for a component that isn't a container.
+      invariant(
+        Relay.isContainer(component),
+        'relay-nested-routes: Route with queries specifies component `%s` ' +
+        'that is not a Relay container.',
+        component.displayName || component.name
+      );
+
+      const routeParams = getParamsForRoute({route, params, location});
+      relayRoute.params = Object.assign(relayRoute.params, routeParams);
+      // TODO: Wipe the entry in this.data on param change.
+
+      Object.keys(queries).forEach(queryName => {
+        const query = queries[queryName];
+        const uniqueQueryName = this.getUniqueQueryName(query, queryName);
+
+        relayRoute.queries[uniqueQueryName] =
+          () => query(component, routeParams);
+        fragmentSpecs[uniqueQueryName] = {component, queryName};
+      });
+    });
+
+    relayRoute.name =
+      ['aggregated', ...Object.keys(relayRoute.queries)].join('-');
+
+    // RootContainer uses referential equality to check for route change, so
+    // replace the route object entirely.
+    this.route = relayRoute;
+    this.fragmentSpecs = fragmentSpecs;
+  }
+
+  getUniqueQueryName(query, queryName) {
+    let uniqueQueryName = this.uniqueQueryNames.get(query);
+    if (uniqueQueryName === undefined) {
+      uniqueQueryName = `$$_${queryName}_${++this.lastQueryIndex}`;
+      this.uniqueQueryNames.set(query, uniqueQueryName);
+    }
+
+    return uniqueQueryName;
+  }
+
+  updateData(data) {
+    this.data = data;
+  }
+
+  getData(queries) {
+    const data = {};
+    for (const queryName of Object.keys(queries)) {
+      const query = queries[queryName];
+      const uniqueQueryName = this.getUniqueQueryName(query, queryName);
+
+      const fragmentPointer = this.data[uniqueQueryName];
+      if (!fragmentPointer) {
+        return null;
+      }
+
+      data[queryName] = fragmentPointer;
+    }
+
+    return data;
+  }
+
+  getFragmentNames() {
```

Comment:
In some sense it does so trivially, though. The only reason we even need to implement `getFragmentNames` there is for https://github.com/facebook/relay/blob/v0.1.1/src/container/getRelayQueries.js#L50, and the fragment name isn't even used anywhere.

In fact it's weird to me that the code there doesn't just run all the queries on the route.

Otherwise if you specify `renderFetched`, `RootContainer` has no actual dependency on the specified `Component` at all, as far as I can tell.